### DISCUSSION TITLE
How is the program executed if there are two `resume`?

### DISCUSSION THREAD

**Discussion by bitbegin** [koka-lang/koka#358] [General]
There are two `resume` in effect handler, how to explain the print results?

```
effect<a> ctl get() : a

fun get-and-print() : <get<int>, console> ()
  println("call 1");
  val m : int = get()
  println("call 2, m: " ++ m.show());
  val n : int = get()
  println("call 3, n: " ++ n.show());
  val l : int = get()
  println("call 4, l: " ++ l.show());
  println("m: " ++ m.show() ++ " n: " ++ n.show() ++ " l: " ++ l.show() ++ "\n")

fun main()
  with ctl get()
    println("get 1");
    resume(10)
    println("get 2");
    resume(20)
    println("get 3");
  get-and-print()

```

output:

```
call 1
get 1
call 2, m: 10
get 1
call 3, n: 10
get 1
call 4, l: 10
m: 10 n: 10 l: 10

get 2
call 4, l: 20
m: 10 n: 10 l: 20

get 3
get 2
call 3, n: 20
get 1
call 4, l: 10
m: 10 n: 20 l: 10

get 2
call 4, l: 20
m: 10 n: 20 l: 20

get 3
get 3
get 2
call 2, m: 20
get 1
call 3, n: 10
get 1
call 4, l: 10
m: 20 n: 10 l: 10

get 2
call 4, l: 20
m: 20 n: 10 l: 20

get 3
get 2
call 3, n: 20
get 1
call 4, l: 10
m: 20 n: 20 l: 10

get 2
call 4, l: 20
m: 20 n: 20 l: 20

get 3
get 3
get 3
```

**Comment by anfelor** [comment]
Hi @bitbegin,

The `resume` keyword resumes the program at the point from the point where the `ctl` operation was called. The surprising feature of Koka here is that you can resume the program multiple times. In the first block of the output, Koka resumes the program from all `get` operations with the result `10`. Then, in the second block, it evaluates the last `get` further and resumes the last `get` with output `20`. And so forth: As you have two `resume`s in three `get` operations, the last `println` is executed 2^3 = 8 times. You can read more about this concept at: https://koka-lang.github.io/koka/doc/book.html#why-handlers
